Z3/src/SimpleSMT/Solver/Z3.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE EmptyDataDecls #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+module SimpleSMT.Solver.Z3
+  ( Z3(..)
+  , new
+  , free
+  , with
+  , toBackend
+  ) where
+
+import qualified SimpleSMT.Solver as Solver
+
+import qualified Data.ByteString.Char8 as BS
+import qualified Data.ByteString.Lazy.Char8 as LBS
+import Data.IORef (newIORef, modifyIORef)
+import qualified Data.Map as M
+import Foreign.Ptr (Ptr)
+import Foreign.ForeignPtr (ForeignPtr, newForeignPtr, finalizeForeignPtr)
+import qualified Language.C.Inline as C
+import qualified Language.C.Inline.Unsafe as CU
+import qualified Language.C.Inline.Context as C
+import qualified Language.C.Types as C
+
+data LogicalContext
+
+C.context
+  (C.baseCtx <>
+   C.fptrCtx <>
+   C.bsCtx <>
+   mempty
+     { C.ctxTypesTable =
+         M.singleton (C.TypeName "Z3_context") [t|Ptr LogicalContext|]
+     })
+C.include "z3.h"
+
+data Z3 = Z3
+    { context :: ForeignPtr LogicalContext
+    -- ^ A black-box representing the internal state of the solver.
+    }
+
+-- | Create a new solver instance.
+new :: IO Z3
+new = do
+  let ctxFinalizer =
+        [C.funPtr| void free_context(Z3_context ctx) {
+                 Z3_del_context(ctx);
+                 } |]
+  ctx <-
+    newForeignPtr ctxFinalizer
+    =<< [CU.block| Z3_context {
+                 Z3_config cfg = Z3_mk_config();
+                 Z3_context ctx = Z3_mk_context(cfg);
+                 Z3_del_config(cfg);
+                 return ctx;
+                 } |]
+  return $ Z3 ctx
+
+-- | Release the resources associated with a Z3 instance.
+free :: Z3 -> IO ()
+free = finalizeForeignPtr . context
+
+-- | Create a Z3 instance, use it to run a computation and release its resources.
+with :: (Z3 -> IO a) -> IO a
+with todo = do
+  z3 <- new
+  result <- todo z3
+  free z3
+  return result
+
+-- | Create a solver backend out of a Z3 instance.
+toBackend :: Z3 -> IO Solver.Backend
+toBackend z3 = do
+  resp <- newIORef mempty
+  return $
+    flip Solver.Backend resp $ \cmd -> do
+      let ctx = context z3
+      -- Z3 requires null-terminated cstrings
+      -- appending the null character performs a memcpy so is inefficient
+      -- TODO a better solution would be to do this on the bytestring-build before it
+      -- is evaluated to a lazy bytestring
```

Comment:
This corresponds to changing the interface of `sendSolver` to demand a NUL-terminated bytestring.

Another solution would be to have `sendSolver` receive a Builder instead of a bytestring, in which case appending a null would be cheaper.

I think that an extra `memcpy` would happen in `toStrict` as a result of appending a NUL character if `cmd` contains at most one chunk. Otherwise, a memcpy would occur regardless of whether the NUL character is appended.

Reply:
> This corresponds to changing the interface of sendSolver to demand a NUL-terminated bytestring.

Yeah I didn't want to do that because this is a Z3-specific problem so it doesn't make sense to enforce that to all builders (this would force the `Process` builder to drop the last character for example).

> Another solution would be to have sendSolver receive a Builder instead of a bytestring, in which case appending a null would be cheaper.

That's what I was thinking of, especially since it also allows backends to specify their own building strategy (`safeStrategy` or `untrimmedStrategy`). But this means both single commands and batches of commands need to be built with the same chunk size (unless the chunk sizes are also passed as arguments, but this would be awkward in the API).

Another solution would be to add the final character (and also the strategy for that matter) as a field in the `Backend` datatype.

> I think that an extra memcpy would happen in toStrict as a result of appending a NUL character if cmd contains at most one chunk. Otherwise, a memcpy would occur regardless of whether the NUL character is appended.

Yeah this is definitely not a pressing matter hence why I'm leaving this as-is at least for now.

Reply:
Havin `send` receive a `Builder` would also have the advantage that the `Process` backend could then directly use the more efficient `hPutBuilder`.

Reply:
> But this means both single commands and batches of commands need to be built with the same chunk size (unless the chunk sizes are also passed as arguments, but this would be awkward in the API).

I think I don't understand what is the need to use different chunk sizes in these scenarios.

Reply:
It's just a matter of optimization:

> A sub-performing case of [toLazyByteString](https://hackage.haskell.org/package/bytestring-0.11.3.1/docs/Data-ByteString-Builder.html#v:toLazyByteString) is executing short (<128 bytes) [Builder](https://hackage.haskell.org/package/bytestring-0.11.3.1/docs/Data-ByteString-Builder.html#t:Builder)s. In this case, the allocation overhead for the first 4kb buffer and the trimming cost dominate the cost of executing the [Builder](https://hackage.haskell.org/package/bytestring-0.11.3.1/docs/Data-ByteString-Builder.html#t:Builder).

IIRC it made a noticeable difference in Pirouette. But if all backends use the `untrimmedStrategy` then that cost is alleviated.

Reply:
https://github.com/tweag/simple-smt/pull/1/commits/3355b1227decba76cda050c6b23b5ec74ad16605 moves the evaluation of the builders inside the backends.